url-updater lacks proper error handling

Our github action to update the urls (#1035) still failed:
https://github.com/devonfw/ide/actions/runs/4815546090/jobs/8574334983
This is the error we now get:
```
java.lang.IllegalStateException: Failed to read body of download https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.612/dotnet-sdk-2.1.612-linux-x64.tar.gz
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doGenerateChecksumFromUrl (AbstractUrlUpdater.java:207)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doGenerateChecksum (AbstractUrlUpdater.java:219)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doUpdateVersion (AbstractUrlUpdater.java:175)
    at com.devonfw.tools.ide.url.updater.dotnet.DotNetUrlUpdater.updateVersion (DotNetUrlUpdater.java:31)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.update (AbstractUrlUpdater.java:295)
    at com.devonfw.tools.ide.url.updater.UpdateManager.updateAll (UpdateManager.java:64)
    at com.devonfw.tools.ide.url.UpdateInitiator.main (UpdateInitiator.java:27)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:829)
Caused by: java.io.IOException: closed
    at jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream.current (ResponseSubscribers.java:352)
    at jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream.read (ResponseSubscribers.java:403)
    at java.io.InputStream.read (InputStream.java:205)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doGenerateChecksumFromUrl (AbstractUrlUpdater.java:200)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doGenerateChecksum (AbstractUrlUpdater.java:219)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doUpdateVersion (AbstractUrlUpdater.java:175)
    at com.devonfw.tools.ide.url.updater.dotnet.DotNetUrlUpdater.updateVersion (DotNetUrlUpdater.java:31)
    at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.update (AbstractUrlUpdater.java:295)
    at com.devonfw.tools.ide.url.updater.UpdateManager.updateAll (UpdateManager.java:64)
    at com.devonfw.tools.ide.url.UpdateInitiator.main (UpdateInitiator.java:27)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:829)
Caused by: java.io.IOException: /10.1.0.25:51998: GOAWAY received
    at jdk.internal.net.http.Http2Connection.handleGoAway (Http2Connection.java:999)
   ...
```

As it seems the server (`dotnetcli.azureedge.net`) got too much load from us and refused further requests via a `GOAWAY`.
The actual problem is that url-updater does not implement a proper error handling (as suggested in my review comments but not implemented). So we have to fix this in order to finally populate our `ide-urls` repository.

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/model/folder/AbstractUrlFolder.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/model/folder/AbstractUrlFolder.java
@@ -3,9 +3,9 @@ package com.devonfw.tools.ide.url.model.folder;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Stream;
@@ -22,7 +22,11 @@ import com.devonfw.tools.ide.url.model.UrlArtifactWithParent;
 public abstract class AbstractUrlFolder<C extends UrlArtifactWithParent<?>> extends AbstractUrlArtifact
     implements UrlFolder<C> {
 
-  private final Map<String, C> children;
+  private final Map<String, C> childMap;
+
+  private final Set<String> childNames;
+
+  private final Collection<C> children;
 
   /**
    * The constructor.
@@ -33,31 +37,33 @@ public abstract class AbstractUrlFolder<C extends UrlArtifactWithParent<?>> exte
   public AbstractUrlFolder(Path path, String name) {
 
     super(path, name);
-    this.children = new HashMap<>();
+    this.childMap = new HashMap<>();
+    this.childNames = Collections.unmodifiableSet(this.childMap.keySet());
+    this.children = Collections.unmodifiableCollection(this.childMap.values());
   }
 
   @Override
   public int getChildCount() {
 
-    return this.children.size();
+    return this.childMap.size();
   }
 
   @Override
   public C getChild(String name) {
 
-    return this.children.get(name);
+    return this.childMap.get(name);
   }
 
   @Override
   public C getOrCreateChild(String name) {
 
-    return this.children.computeIfAbsent(name, p -> newChild(name));
+    return this.childMap.computeIfAbsent(name, p -> newChild(name));
   }
 
   @Override
-  public List<C> getChildren() {
+  public Collection<C> getChildren() {
 
-    return new ArrayList<>(this.children.values());
+    return this.children;
   }
 
   /**
@@ -73,13 +79,10 @@ public abstract class AbstractUrlFolder<C extends UrlArtifactWithParent<?>> exte
 
   /**
    * @return the {@link Set} with all {@link #getName() names} of the children.
-   * @deprecated method name does not make sense here and the method exposes internal mutable structures allowing to
-   *             remove or add items to the internal map bypassing the API.
    */
-  @Deprecated
-  public Set<String> getAllVersions() {
+  public Set<String> getChildNames() {
 
-    return this.children.keySet();
+    return this.childNames;
   }
 
   @Override
@@ -115,7 +118,7 @@ public abstract class AbstractUrlFolder<C extends UrlArtifactWithParent<?>> exte
   @Override
   public void save() {
 
-    for (C child : this.children.values()) {
+    for (C child : this.childMap.values()) {
       ((AbstractUrlArtifact) child).save();
     }
   }

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
@@ -6,12 +6,10 @@ import java.net.URI;
 import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.time.Duration;
+import java.time.Instant;
 import java.util.Collection;
 import java.util.Set;
 
@@ -23,7 +21,9 @@ import com.devonfw.tools.ide.common.OperatingSystem;
 import com.devonfw.tools.ide.common.SystemArchitecture;
 import com.devonfw.tools.ide.url.model.file.UrlChecksum;
 import com.devonfw.tools.ide.url.model.file.UrlDownloadFile;
+import com.devonfw.tools.ide.url.model.file.UrlFile;
 import com.devonfw.tools.ide.url.model.file.UrlStatusFile;
+import com.devonfw.tools.ide.url.model.file.json.StatusJson;
 import com.devonfw.tools.ide.url.model.file.json.UrlStatus;
 import com.devonfw.tools.ide.url.model.file.json.UrlStatusState;
 import com.devonfw.tools.ide.url.model.folder.UrlEdition;
@@ -37,6 +37,8 @@ import com.devonfw.tools.ide.util.HexUtil;
  * updating tool versions, and checking if download URLs work.
  */
 public abstract class AbstractUrlUpdater implements UrlUpdater {
+
+  private static final Duration TWO_DAYS = Duration.ofDays(2);
 
   /** {@link OperatingSystem#WINDOWS}. */
   protected static final OperatingSystem WINDOWS = OperatingSystem.WINDOWS;
@@ -145,43 +147,38 @@ public abstract class AbstractUrlUpdater implements UrlUpdater {
    * Updates a tool version with the given arguments.
    *
    * @param urlVersion the UrlVersion instance to update.
-   * @param downloadUrl the URL of the download for the tool.
+   * @param url the URL of the download for the tool.
    * @param os the optional {@link OperatingSystem}.
    * @param architecture the optional {@link SystemArchitecture}.
    * @return {@code true} if the version was successfully updated, {@code false} otherwise.
    */
-  protected boolean doUpdateVersion(UrlVersion urlVersion, String downloadUrl, OperatingSystem os,
+  protected boolean doUpdateVersion(UrlVersion urlVersion, String url, OperatingSystem os,
       SystemArchitecture architecture) {
 
     String version = urlVersion.getName();
-    downloadUrl = downloadUrl.replace("${version}", version);
+    url = url.replace("${version}", version);
     if (os != null) {
-      downloadUrl = downloadUrl.replace("${os}", os.toString());
+      url = url.replace("${os}", os.toString());
     }
     if (architecture != null) {
-      downloadUrl = downloadUrl.replace("${arch}", architecture.toString());
+      url = url.replace("${arch}", architecture.toString());
     }
-    downloadUrl = downloadUrl.replace("${edition}", getEdition());
-    UrlRequestResult result = doCheckIfDownloadUrlWorks(downloadUrl);
+    url = url.replace("${edition}", getEdition());
+    UrlRequestResult result = doCheckIfDownloadUrlWorks(url);
     if (result.isSuccess()) {
       UrlDownloadFile urlDownloadFile = urlVersion.getOrCreateUrls(os, architecture);
-      urlDownloadFile.addUrl(downloadUrl);
-      doCreateOrRefreshStatusJson(result, urlVersion, downloadUrl);
-      urlVersion.save();
-
+      urlDownloadFile.addUrl(url);
+      doUpdateStatusJson(result, urlVersion, url);
       // generate checksum of download file
       if (result.getHttpStatusCode() == 200) {
         UrlChecksum urlChecksum = urlVersion.getOrCreateChecksum(urlDownloadFile.getName());
-        doGenerateChecksum(downloadUrl, urlChecksum);
+        doGenerateChecksum(url, urlChecksum);
       }
+      urlVersion.save();
       return true;
     } else {
-      // check if folder of urlVersion exists
-      Path folderPath = Paths.get(urlVersion.getPath().toString());
-      if (Files.exists(folderPath) && Files.isDirectory(folderPath)) {
-        doCreateOrRefreshStatusJson(result, urlVersion, downloadUrl);
-        urlVersion.save();
-      }
+      logger.error("For tool {} and version {} the download of {} failed with status {}", getToolWithEdition(), version,
+          url, result.getHttpStatusCode());
       return false;
     }
   }
@@ -261,18 +258,21 @@ public abstract class AbstractUrlUpdater implements UrlUpdater {
    * @param result the {@link UrlRequestResult} instance indicating whether the download URL works.
    * @param urlVersion the UrlVersion instance to create or refresh the status JSON file for.
    */
-  private void doCreateOrRefreshStatusJson(UrlRequestResult result, UrlVersion urlVersion, String downloadUrl) {
+  private void doUpdateStatusJson(UrlRequestResult result, UrlVersion urlVersion, String url) {
 
     UrlStatusFile urlStatusFile = urlVersion.getOrCreateStatus();
-    UrlStatus status = urlStatusFile.getStatusJson().getOrCreateUrlStatus(downloadUrl);
+    UrlStatus status = urlStatusFile.getStatusJson().getOrCreateUrlStatus(url);
     if (result.isSuccess()) {
       status.setSuccess(new UrlStatusState());
+      logger.info("Successfully verified download URL {}.", url);
     } else if (result.isFailure()) {
       UrlStatusState error = new UrlStatusState();
-      error.setCode(Integer.valueOf(result.getHttpStatusCode()));
+      Integer code = Integer.valueOf(result.getHttpStatusCode());
+      error.setCode(code);
       // String message = result.getHttpStatusCode() + " " + result.getUrl();
       // error.setMessage(message);
       status.setError(error);
+      logger.warn("Download verification failed with status code {} for URL {}.", code, url);
     }
   }
 
@@ -288,12 +288,17 @@ public abstract class AbstractUrlUpdater implements UrlUpdater {
     UrlEdition edition = tool.getOrCreateChild(getEdition());
     updateExistingVersions(edition);
     Set<String> versions = getVersions();
+    String toolWithEdition = getToolWithEdition();
+    logger.info("For tool {} we found the following versions : {}", toolWithEdition, versions);
     for (String version : versions) {
-      version = mapVersion(version);
-      if (version != null && edition.getChild(version) == null && !version.isEmpty()) {
-        UrlVersion urlVersion = edition.getOrCreateChild(version);
-        updateVersion(urlVersion);
-        urlVersion.save();
+      if (edition.getChild(version) == null) {
+        try {
+          UrlVersion urlVersion = edition.getOrCreateChild(version);
+          updateVersion(urlVersion);
+          urlVersion.save();
+        } catch (Exception e) {
+          logger.error("For tool {} we failed to add version {}.", toolWithEdition, version, e);
+        }
       }
     }
   }
@@ -305,19 +310,55 @@ public abstract class AbstractUrlUpdater implements UrlUpdater {
    */
   protected void updateExistingVersions(UrlEdition edition) {
 
-    Set<String> existingVersions = edition.getAllVersions();
+    Set<String> existingVersions = edition.getChildNames();
     for (String version : existingVersions) {
       UrlVersion urlVersion = edition.getChild(version);
       if (urlVersion != null) {
         UrlStatusFile urlStatusFile = urlVersion.getOrCreateStatus();
-        logger.info("Getting or creating Status for version {}", version);
-        if (urlStatusFile.getStatusJson().isManual()) {
-          logger.info("Version {} is manual, skipping update", version);
-          continue;
+        StatusJson statusJson = urlStatusFile.getStatusJson();
+        if (statusJson.isManual()) {
+          logger.info("For tool {} the version {} is set to manual, hence skipping update", getToolWithEdition(),
+              version);
+        } else {
+          updateExistingVersion(version, urlVersion, statusJson, urlStatusFile);
+          urlVersion.save();
         }
-        updateVersion(urlVersion);
-        urlVersion.save();
       }
+    }
+  }
+
+  private void updateExistingVersion(String version, UrlVersion urlVersion, StatusJson statusJson,
+      UrlStatusFile urlStatusFile) {
+
+    boolean modified = false;
+    String toolWithEdition = getToolWithEdition();
+    Instant now = Instant.now();
+    for (UrlFile child : urlVersion.getChildren()) {
+      if (child instanceof UrlDownloadFile) {
+        Set<String> urls = ((UrlDownloadFile) child).getUrls();
+        for (String url : urls) {
+          UrlStatus urlStatus = statusJson.getOrCreateUrlStatus(url);
+          UrlStatusState success = urlStatus.getSuccess();
+          if (success != null) {
+            Instant timestamp = success.getTimestamp();
+            if (timestamp != null) {
+              Duration duration = Duration.between(timestamp, now);
+              if (duration.compareTo(TWO_DAYS) <= 0) {
+                logger.debug("For tool {} the URL {} has already been checked recently on {}", toolWithEdition, url,
+                    timestamp);
+                continue;
+              }
+            }
+          }
+          UrlRequestResult result = doCheckIfDownloadUrlWorks(url);
+          doUpdateStatusJson(result, urlVersion, url);
+          modified = true;
+        }
+      }
+    }
+    if (modified) {
+      urlStatusFile.setStatusJson(statusJson); // hack to set modified (better solution welcome)
+      urlStatusFile.save();
     }
   }
 
@@ -344,8 +385,9 @@ public abstract class AbstractUrlUpdater implements UrlUpdater {
   protected final void addVersion(String version, Collection<String> versions) {
 
     String mappedVersion = mapVersion(version);
-    if (mappedVersion == null) {
+    if ((mappedVersion == null) || mappedVersion.isBlank()) {
       logger.debug("Filtered version {}", version);
+      return;
     } else if (!mappedVersion.equals(version)) {
       logger.debug("Mapped version {} to {}", version, mappedVersion);
     }

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/GithubUrlUpdater.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/GithubUrlUpdater.java
@@ -27,12 +27,18 @@ public abstract class GithubUrlUpdater extends JsonUrlUpdater<GithubTags> {
 
     for (GithubTag item : jsonItem) {
       String version = item.getRef().replace("refs/tags/", "");
-      if (version.contains("alpha") || version.contains("beta") || version.contains("dev") || version.contains("rc")
-          || version.contains("snapshot") || version.contains("preview") || version.equals("")) {
-        continue;
-      }
       addVersion(version, versions);
     }
+  }
+
+  @Override
+  protected String mapVersion(String version) {
+
+    if (version.contains("alpha") || version.contains("beta") || version.contains("dev") || version.contains("rc")
+        || version.contains("snapshot") || version.contains("preview")) {
+      return null;
+    }
+    return super.mapVersion(version);
   }
 
   /**

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/JsonUrlUpdater.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/JsonUrlUpdater.java
@@ -4,9 +4,6 @@ import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import com.devonfw.tools.ide.common.JsonObject;
 import com.devonfw.tools.ide.json.mapping.JsonMapping;
 import com.devonfw.tools.ide.url.model.folder.UrlEdition;
@@ -21,8 +18,6 @@ import com.fasterxml.jackson.databind.ObjectMapper;
  */
 public abstract class JsonUrlUpdater<J extends JsonObject> extends AbstractUrlUpdater {
 
-  private final Logger logger = LoggerFactory.getLogger(JsonUrlUpdater.class.getName());
-
   private static final ObjectMapper MAPPER = JsonMapping.create();
 
   @Override
@@ -34,16 +29,25 @@ public abstract class JsonUrlUpdater<J extends JsonObject> extends AbstractUrlUp
       String response = doGetResponseBodyAsString(url);
       J jsonObject = MAPPER.readValue(response, getJsonObjectType());
       collectVersionsFromJson(jsonObject, versions);
-      this.logger.info("Found following versions for tool {} in JSON {}", getToolWithEdition(), versions);
     } catch (Exception e) {
       throw new IllegalStateException("Error while getting versions from JSON API " + url, e);
     }
     return versions;
   }
 
+  /**
+   * @return the URL of the JSON API to get the versions from.
+   */
   protected abstract String doGetVersionUrl();
 
+  /**
+   * @return the {@link Class} reflecting the Java object the JSON shall be mapped to.
+   */
   protected abstract Class<J> getJsonObjectType();
 
+  /**
+   * @param jsonItem the Java object parsed from the JSON holding the information of a version.
+   * @param versions the versions where to {@link #addVersion(String, Collection) add the version to}.
+   */
   protected abstract void collectVersionsFromJson(J jsonItem, Collection<String> versions);
 }

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/MavenBasedUrlUpdater.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/MavenBasedUrlUpdater.java
@@ -4,9 +4,6 @@ import java.io.IOException;
 import java.util.HashSet;
 import java.util.Set;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import com.devonfw.tools.ide.maven.MavenMetadata;
 import com.devonfw.tools.ide.url.model.folder.UrlVersion;
 import com.fasterxml.jackson.dataformat.xml.XmlMapper;
@@ -15,7 +12,8 @@ import com.fasterxml.jackson.dataformat.xml.XmlMapper;
  * The MvnCrawler class is an abstract class that provides functionality for crawling Maven repositories.
  */
 public abstract class MavenBasedUrlUpdater extends AbstractUrlUpdater {
-  private final static Logger logger = LoggerFactory.getLogger(MavenBasedUrlUpdater.class.getName());
+
+  private static final String MAVEN_METADATA_XML = "maven-metadata.xml";
 
   private final String mavenBaseRepoUrl;
 
@@ -51,7 +49,7 @@ public abstract class MavenBasedUrlUpdater extends AbstractUrlUpdater {
   @Override
   protected Set<String> getVersions() {
 
-    return doGetVersionsFromMavenApi(this.mavenBaseRepoUrl + "maven-metadata.xml");
+    return doGetVersionsFromMavenApi(this.mavenBaseRepoUrl + MAVEN_METADATA_XML);
   }
 
   @Override
@@ -79,9 +77,8 @@ public abstract class MavenBasedUrlUpdater extends AbstractUrlUpdater {
         addVersion(version, versions);
       }
     } catch (IOException e) {
-      logger.error("URLStatusError while getting javaJsonVersions", e);
+      throw new IllegalStateException("Failed to get version from " + url, e);
     }
-    logger.info("Found  javaJsonVersions : {}", versions);
     return versions;
   }
 

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/UpdateManager.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/UpdateManager.java
@@ -4,6 +4,9 @@ import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.List;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.devonfw.tools.ide.url.model.folder.UrlRepository;
 import com.devonfw.tools.ide.url.updater.aws.AwsUrlUpdater;
 import com.devonfw.tools.ide.url.updater.az.AzureUrlUpdater;
@@ -38,30 +41,41 @@ import com.devonfw.tools.ide.url.updater.vscode.VSCodeUrlUpdater;
 
 /**
  * The {@code UpdateManager} class manages the update process for various tools by using a list of
- * {@link AbstractUrlUpdater}s to update the {@link UrlRepository}. The list of {@link AbstractUrlUpdater}s contains crawlers
- * for different tools and services, To use the UpdateManager, simply create an instance with the path to the repository
- * as a parameter and call the {@link #updateAll()} method.
+ * {@link AbstractUrlUpdater}s to update the {@link UrlRepository}. The list of {@link AbstractUrlUpdater}s contains
+ * crawlers for different tools and services, To use the UpdateManager, simply create an instance with the path to the
+ * repository as a parameter and call the {@link #updateAll()} method.
  */
 public class UpdateManager {
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractUrlUpdater.class);
+
   private final UrlRepository urlRepository;
 
-  private final List<AbstractUrlUpdater> crawlers = Arrays.asList(new AwsUrlUpdater(), new AzureUrlUpdater(),
+  private final List<AbstractUrlUpdater> updaters = Arrays.asList(new AwsUrlUpdater(), new AzureUrlUpdater(),
       new CobigenUrlUpdater(), new DotNetUrlUpdater(), new DockerDesktopUrlUpdater(), new EclipseCppUrlUpdater(),
-      new EclipseJavaUrlUpdater(), new GcViewerUrlUpdater(), new GhUrlUpdater(), new GraalVmUrlUpdater(), new GradleUrlUpdater(),
-      new HelmUrlUpdater(), new IntellijUltimateUrlUpdater(), new IntellijCommunityUrlUpdater(), new JavaUrlUpdater(),
-      new JenkinsUrlUpdater(), new KotlincUrlUpdater(), new KotlincNativeUrlUpdater(), new LazyDockerUrlUpdater(), new MvnUrlUpdater(),
-      new NodeUrlUpdater(), new NpmUrlUpdater(), new OcUrlUpdater(), new PipUrlUpdater(), new PythonUrlUpdater(), new QuarkusUrlUpdater(),
-      new DockerRancherDesktopUrlUpdater(), new SonarUrlUpdater(), new TerraformUrlUpdater(), new VSCodeUrlUpdater());
+      new EclipseJavaUrlUpdater(), new GcViewerUrlUpdater(), new GhUrlUpdater(), new GraalVmUrlUpdater(),
+      new GradleUrlUpdater(), new HelmUrlUpdater(), new IntellijUltimateUrlUpdater(), new IntellijCommunityUrlUpdater(),
+      new JavaUrlUpdater(), new JenkinsUrlUpdater(), new KotlincUrlUpdater(), new KotlincNativeUrlUpdater(),
+      new LazyDockerUrlUpdater(), new MvnUrlUpdater(), new NodeUrlUpdater(), new NpmUrlUpdater(), new OcUrlUpdater(),
+      new PipUrlUpdater(), new PythonUrlUpdater(), new QuarkusUrlUpdater(), new DockerRancherDesktopUrlUpdater(),
+      new SonarUrlUpdater(), new TerraformUrlUpdater(), new VSCodeUrlUpdater());
 
   public UpdateManager(Path pathToRepository) {
 
     this.urlRepository = UrlRepository.load(pathToRepository);
   }
 
+  /**
+   * Updates {@code ide-urls} for all tools their editions and all found versions.
+   */
   public void updateAll() {
 
-    for (AbstractUrlUpdater crawler : crawlers) {
-      crawler.update(urlRepository);
+    for (AbstractUrlUpdater updater : this.updaters) {
+      try {
+        updater.update(this.urlRepository);
+      } catch (Exception e) {
+        logger.error("Failed to update {}", updater.getToolWithEdition(), e);
+      }
     }
   }
 

--- a/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/WebsiteUrlUpdater.java
+++ b/url-updater/src/main/java/com/devonfw/tools/ide/url/updater/WebsiteUrlUpdater.java
@@ -2,8 +2,6 @@ package com.devonfw.tools.ide.url.updater;
 
 import java.util.HashSet;
 import java.util.Set;
-import java.util.logging.Logger;
-import java.util.regex.MatchResult;
 import java.util.regex.Pattern;
 
 /**
@@ -16,7 +14,6 @@ import java.util.regex.Pattern;
  * version pattern and URL of the website to crawl, respectively.
  */
 public abstract class WebsiteUrlUpdater extends AbstractUrlUpdater {
-  private final Logger logger = Logger.getLogger(WebsiteUrlUpdater.class.getName());
 
   /**
    * Retrieves the available versions by performing a GET request on the version URL and extracting the version numbers
@@ -42,11 +39,9 @@ public abstract class WebsiteUrlUpdater extends AbstractUrlUpdater {
     Set<String> versions = new HashSet<>();
     var matcher = getVersionPattern().matcher(htmlBody);
     while (matcher.find()) {
-      MatchResult result = matcher.toMatchResult();
-      String match = result.group();
-      versions.add(match);
+      String version = matcher.group();
+      addVersion(version, versions);
     }
-    logger.info("Found  versions : " + versions);
     return versions;
   }
 

